packages/beacon-node/src/chain/blocks/verifyBlocksDataAvailability.ts

```diff
@@ -0,0 +1,126 @@
+  }
+
+  const availableTime = blocks[blocks.length - 1].type === BlockInputType.blobsPromise ? Date.now() : seenTime;
+  if (blocks.length === 1 && opts.seenTimestampSec !== undefined && blocks[0].type !== BlockInputType.preDeneb) {
+    const recvToAvailableTime = availableTime / 1000 - opts.seenTimestampSec;
+    const numBlobs = (blocks[0].block as deneb.SignedBeaconBlock).message.body.blobKzgCommitments.length;
+
+    chain.metrics?.gossipBlock.receivedToBlobsAvailabityTime.observe({numBlobs}, recvToAvailableTime);
+    chain.logger.verbose("Verified blobs availability", {
+      slot: blocks[0].block.message.slot,
+      recvToAvailableTime,
+      type: blocks[0].type,
+    });
+  }
+
+  return {dataAvailabilityStatuses, availableTime};
+}
+
+async function maybeValidateBlobs(
+  chain: {config: ChainForkConfig; genesisTime: UintNum64},
+  blockInput: BlockInput,
+  opts: ImportBlockOpts
+): Promise<DataAvailableStatus> {
+  switch (blockInput.type) {
+    case BlockInputType.preDeneb:
+      return DataAvailableStatus.preDeneb;
+
+    case BlockInputType.postDeneb:
+      if (opts.validBlobSidecars === BlobSidecarValidation.Full) {
+        return DataAvailableStatus.available;
+      }
+
+    // eslint-disable-next-line no-fallthrough
+    case BlockInputType.blobsPromise: {
+      // run full validation
+      const {block} = blockInput;
+      const blockSlot = block.message.slot;
+
+      const blobsData =
+        blockInput.type === BlockInputType.postDeneb
+          ? blockInput
+          : await raceWithCutoff(chain, blockInput, blockInput.availabilityPromise);
+      const {blobs} = blobsData;
+
+      const {blobKzgCommitments} = (block as deneb.SignedBeaconBlock).message.body;
+      const beaconBlockRoot = chain.config.getForkTypes(blockSlot).BeaconBlock.hashTreeRoot(block.message);
+
+      // if the blob siddecars have been individually verified then we can skip kzg proof check
+      // but other checks to match blobs with block data still need to be performed
+      const skipProofsCheck = opts.validBlobSidecars === BlobSidecarValidation.Individual;
+      validateBlobSidecars(blockSlot, beaconBlockRoot, blobKzgCommitments, blobs, {skipProofsCheck});
+
+      return DataAvailableStatus.available;
+    }
+  }
+}
+
+/**
+ * Wait for blobs to become available with a cutoff time. If fails then throw DATA_UNAVAILABLE error
+ * which may try unknownblock/blobs fill (by root).
+ */
+async function raceWithCutoff<T>(
+  chain: {config: ChainForkConfig; genesisTime: UintNum64},
+  blockInput: BlockInput,
+  availabilityPromise: Promise<T>
+): Promise<T> {
+  const {block} = blockInput;
+  const blockSlot = block.message.slot;
+
+  const cutoffTime = Math.max(
+    computeTimeAtSlot(chain.config, blockSlot, chain.genesisTime) * 1000 + BLOB_AVAILABILITY_TIMEOUT - Date.now(),
+    0
+  );
+  const cutoffTimeout = new Promise((_resolve, reject) => setTimeout(reject, cutoffTime));
```

Comment:
maybe reject with a reason, otherwise it'll be hard to debug later

Reply:
a proper error is thrown at L122